include/spinlock.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _SPINLOCK_H
+#define _SPINLOCK_H
+
+#include <atomic.h>
+
+struct k_spinlock_key {
+	int key;
+};
+
+typedef struct k_spinlock_key k_spinlock_key_t;
+
+struct k_spinlock {
+#ifdef CONFIG_SMP
+	atomic_t locked;
+#ifdef CONFIG_DEBUG
+	int saved_key;
+#endif
+#endif
+};
+
+static inline k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
+{
+	k_spinlock_key_t k;
+
+	/* Note that we need to use the underlying arch-specific lock
+	 * implementation.  The "irq_lock()" API in SMP context is
+	 * actually a wrapper for a global spinlock!
+	 */
+	k.key = _arch_irq_lock();
+
+#ifdef CONFIG_SMP
+# ifdef CONFIG_DEBUG
+	l->saved_key = k.key;
+# endif
+	while (!atomic_cas(&l->locked, 0, 1)) {
+	}
+#endif
+
+	return k;
+}
+
+static inline void k_spin_unlock(struct k_spinlock *l, k_spinlock_key_t key)
+{
+#ifdef CONFIG_SMP
+# ifdef CONFIG_DEBUG
+	/* This doesn't attempt to catch all mismatches, just spots
+	 * where the arch state register shows a difference.  Should
+	 * add a nesting count or something...
+	 */
+	__ASSERT(l->saved_key == key.key, "Mismatched spin lock/unlock");
+# endif
+	/* Strictly we don't need atomic_clear() here (which is an
+	 * exchange operation that returns the old value).  We are always
+	 * setting a zero and (because we hold the lock) know the existing
+	 * state won't change due to a race.  But some architectures need
+	 * a memory barrier when used like this, and we don't have a
+	 * Zephyr framework for that.
```

Comment:
We have a memory barrier macro which should enforce ordering of memory access:

```
include/toolchain/gcc.h
331-#endif
332-
333:#define compiler_barrier() do { \
334-	__asm__ __volatile__ ("" ::: "memory"); \
335-} while ((0))
```

Or do you need something more like an instruction barrier?

Reply:
That comment was sort of obtuse.  I mean a hardware barrier.  Architectures with relaxed cache coherence rules (none of ours, I'm told that older NUMA systems in the 80's did this) wouldn't expose a C assignment (i.e." locked = 0;") to the cache in such a way that other CPUs would see the lock released, so another could come along and get an inconsistent result.  Using the atomic API guarantees that we're using a correct primitive as long as the architecture does it right.